File: spike/src/language_detector.cpp
#include <algorithm>
#include <cmath>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <regex>
#include <sstream>
#include <string>
#include <vector>

namespace {

constexpr std::size_t kMaxInputChars = 20000;

struct Score {
    int value{0};
    std::vector<std::string> evidence;
};

bool contains(const std::string& text, const std::string& token) {
    return text.find(token) != std::string::npos;
}

bool matches(const std::string& text, const std::string& pattern) {
    return std::regex_search(text, std::regex(pattern));
}

void add_feature(Score& score, int weight, const std::string& label) {
    score.value += weight;
    score.evidence.push_back(label);
}

Score score_cpp(const std::string& code) {
    Score score;
    if (contains(code, "#include")) {
        add_feature(score, 5, "cpp:#include");
    }
    if (matches(code, R"(\bint\s+main\s*\()")) {
        add_feature(score, 5, "cpp:int-main");
    }
    if (contains(code, "std::")) {
        add_feature(score, 4, "cpp:std-namespace");
    }
    if (matches(code, R"(\b(vector|string|map|unordered_map|set)\s*<)")) {
        add_feature(score, 3, "cpp:template-type");
    }
    if (std::count(code.begin(), code.end(), ';') >= 2 &&
        contains(code, "{") && contains(code, "}")) {
        add_feature(score, 2, "cpp:brace-semicolon-structure");
    }
    if (contains(code, "nullptr") ||
        (score.value >= 5 && contains(code, "->"))) {
        add_feature(score, 2, "cpp:pointer-syntax");
    }
    return score;
}

Score score_python(const std::string& code) {
    Score score;
    if (matches(code, R"((^|\n)\s*(async\s+)?def\s+[A-Za-z_]\w*\s*\([^\n]*\)\s*:)")) {
        add_feature(score, 5, "python:def-block");
    }
    if (matches(code, R"((^|\n)\s*(from\s+\S+\s+import|import\s+\S+))")) {
        add_feature(score, 4, "python:import-statement");
    }
    if (contains(code, "if __name__")) {
        add_feature(score, 5, "python:main-guard");
    }
    if (matches(code, R"(\b(None|True|False)\b)")) {
        add_feature(score, 2, "python:built-in-literal");
    }
    if (matches(code, R"((^|\n)\s*(for|while|if|elif|else|try|except|with|class)\b[^\n]*:)")) {
        add_feature(score, 3, "python:colon-block");
    }
    if (score.value >= 3 && contains(code, "\n    ") && contains(code, ":")) {
        add_feature(score, 2, "python:indented-block");
    }
    return score;
}

std::string json_array(const std::vector<std::string>& values) {
    std::ostringstream out;
    out << '[';
    for (std::size_t index = 0; index < values.size(); ++index) {
        if (index > 0) {
            out << ',';
        }
        out << '"' << values[index] << '"';
    }
    out << ']';
    return out.str();
}

bool is_blank(const std::string& text) {
    return std::all_of(text.begin(), text.end(), [](unsigned char value) {
        return std::isspace(value) != 0;
    });
}

}  // namespace

int main(int argc, char* argv[]) {
    if (argc != 2) {
        std::cerr << "usage: language_detector <code-file>\n";
        return 2;
    }

    std::ifstream input(argv[1], std::ios::binary);
    if (!input) {
        std::cerr << "cannot read input file\n";
        return 2;
    }

    std::ostringstream buffer;
    buffer << input.rdbuf();
    const std::string code = buffer.str();

    if (code.empty() || is_blank(code)) {
        std::cout << R"({"error":"empty_code","detected_language":"unknown"})" << '\n';
        return 2;
    }
    if (code.size() > kMaxInputChars) {
        std::cout << R"({"error":"code_too_large","detected_language":"unknown"})" << '\n';
        return 2;
    }

    const Score cpp = score_cpp(code);
    const Score python = score_python(code);
    const int maximum = std::max(cpp.value, python.value);
    const int difference = std::abs(cpp.value - python.value);

    std::string language = "unknown";
    std::vector<std::string> evidence;
    double confidence = std::min(0.79, maximum / 10.0);

    if (maximum >= 5 && difference >= 3) {
        const bool is_cpp = cpp.value > python.value;
        language = is_cpp ? "cpp" : "python";
        evidence = is_cpp ? cpp.evidence : python.evidence;
        confidence = std::min(0.99, 0.75 + difference * 0.02);
    }

    std::cout << std::fixed << std::setprecision(2)
              << "{\"detected_language\":\"" << language
              << "\",\"confidence\":" << confidence
              << ",\"cpp_score\":" << cpp.value
              << ",\"python_score\":" << python.value
              << ",\"detection_evidence\":" << json_array(evidence)
              << "}" << '\n';
    return 0;
}

File: app/detectors/language_detector.py
from __future__ import annotations

import ast
import re
import shutil
import subprocess
import tempfile
from dataclasses import dataclass
from pathlib import Path

from app.domain.contracts import DetectionResult
from app.security.process_limits import limited_syntax_command


@dataclass
class _Score:
    value: int
    evidence: list[str]


def _score_cpp(code: str) -> _Score:
    value = 0
    evidence: list[str] = []

    def add(weight: int, label: str) -> None:
        nonlocal value
        value += weight
        evidence.append(label)

    if "#include" in code:
        add(5, "cpp:#include")
    if re.search(r"\bint\s+main\s*\(", code):
        add(5, "cpp:int-main")
    if "std::" in code:
        add(4, "cpp:std-namespace")
    if re.search(r"\b(vector|string|map|unordered_map|set)\s*<", code):
        add(3, "cpp:template-type")
    if code.count(";") >= 2 and "{" in code and "}" in code:
        add(2, "cpp:brace-semicolon-structure")
    if "nullptr" in code or (value >= 5 and "->" in code):
        add(2, "cpp:pointer-syntax")
    return _Score(value, evidence)


def _score_python(code: str) -> _Score:
    value = 0
    evidence: list[str] = []

    def add(weight: int, label: str) -> None:
        nonlocal value
        value += weight
        evidence.append(label)

    if re.search(r"(^|\n)\s*(async\s+)?def\s+[A-Za-z_]\w*\s*\([^\n]*\)\s*:", code):
        add(5, "python:def-block")
    if re.search(r"(^|\n)\s*(from\s+\S+\s+import|import\s+\S+)", code):
        add(4, "python:import-statement")
    if "if __name__" in code:
        add(5, "python:main-guard")
    if re.search(r"\b(None|True|False)\b", code):
        add(2, "python:built-in-literal")
    if re.search(
        r"(^|\n)\s*(for|while|if|elif|else|try|except|with|class)\b[^\n]*:",
        code,
    ):
        add(3, "python:colon-block")
    if value >= 3 and "\n    " in code and ":" in code:
        add(2, "python:indented-block")
    return _Score(value, evidence)


def _probe_python(code: str) -> str:
    try:
        ast.parse(code)
    except (SyntaxError, ValueError, TypeError):
        return "failed"
    return "passed"


def _probe_cpp(code: str, timeout_seconds: float) -> str:
    compiler = shutil.which("g++")
    if compiler is None:
        return "unavailable"

    try:
        with tempfile.TemporaryDirectory(prefix="code_coach_probe_") as directory:
            source_path = Path(directory) / "input.cpp"
            source_path.write_text(code, encoding="utf-8")
            command = limited_syntax_command(
                [
                    compiler,
                    "-std=c++17",
                    "-x",
                    "c++",
                    "-fsyntax-only",
                    str(source_path),
                ]
            )
            completed = subprocess.run(
                command,
                cwd=directory,
                capture_output=True,
                text=True,
                timeout=timeout_seconds,
                check=False,
            )
    except subprocess.TimeoutExpired:
        return "timeout"
    except OSError:
        return "unavailable"
    return "passed" if completed.returncode == 0 else "failed"


class LanguageDetector:
    def __init__(self, probe_timeout_seconds: float = 3.0):
        self.probe_timeout_seconds = probe_timeout_seconds

    def detect(self, code: str) -> DetectionResult:
        cpp = _score_cpp(code)
        python = _score_python(code)
        cpp_probe = _probe_cpp(code, self.probe_timeout_seconds)
        python_probe = _probe_python(code)

        maximum = max(cpp.value, python.value)
        difference = abs(cpp.value - python.value)
        strong_candidate = maximum >= 5 and difference >= 3

        if strong_candidate:
            language = "cpp" if cpp.value > python.value else "python"
            score = cpp if language == "cpp" else python
            probe = cpp_probe if language == "cpp" else python_probe
            confidence = 0.95 if probe == "passed" else 0.82
            evidence = [*score.evidence, f"{language}_probe:{probe}"]
            return DetectionResult(
                language=language,
                confidence=confidence,
                evidence=evidence,
                cpp_score=cpp.value,
                python_score=python.value,
                cpp_probe=cpp_probe,
                python_probe=python_probe,
                decision_reason="strong_feature_candidate",
            )

        weak_cpp = cpp.value > python.value and cpp.value > 0
        weak_python = python.value > cpp.value and python.value > 0
        if weak_cpp and cpp_probe == "passed" and python_probe == "failed":
            return DetectionResult(
                language="cpp",
                confidence=0.80,
                evidence=[*cpp.evidence, "cpp_probe:passed"],
                cpp_score=cpp.value,
                python_score=python.value,
                cpp_probe=cpp_probe,
                python_probe=python_probe,
                decision_reason="weak_features_with_single_probe",
            )
        if weak_python and python_probe == "passed" and cpp_probe == "failed":
            return DetectionResult(
                language="python",
                confidence=0.80,
                evidence=[*python.evidence, "python_probe:passed"],
                cpp_score=cpp.value,
                python_score=python.value,
                cpp_probe=cpp_probe,
                python_probe=python_probe,
                decision_reason="weak_features_with_single_probe",
            )

        confidence = min(0.79, maximum / 10.0)
        return DetectionResult(
            language="unknown",
            confidence=confidence,
            evidence=[
                f"cpp_score:{cpp.value}",
                f"python_score:{python.value}",
                f"cpp_probe:{cpp_probe}",
                f"python_probe:{python_probe}",
            ],
            cpp_score=cpp.value,
            python_score=python.value,
            cpp_probe=cpp_probe,
            python_probe=python_probe,
            decision_reason="insufficient_or_conflicting_evidence",
        )
